File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/modules/render/FreeCam.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.render;

import net.ccbluex.liquidbounce.event.EventTarget;
import net.ccbluex.liquidbounce.event.PacketEvent;
import net.ccbluex.liquidbounce.event.UpdateEvent;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.utils.MovementUtils;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.FloatValue;
import net.minecraft.client.entity.EntityOtherPlayerMP;
import net.minecraft.network.Packet;
import net.minecraft.network.play.client.C03PacketPlayer;
import net.minecraft.network.play.client.C0BPacketEntityAction;

@ModuleInfo(name = "FreeCam", description = "Allows you to move out of your body.", category = ModuleCategory.RENDER)
public class FreeCam extends Module {
    private final FloatValue speedValue = new FloatValue("Speed", 0.8F, 0.1F, 2F);
    private final BoolValue flyValue = new BoolValue("Fly", true);
    private final BoolValue noClipValue = new BoolValue("NoClip", true);

    private EntityOtherPlayerMP fakePlayer = null;
    private double oldX;
    private double oldY;
    private double oldZ;

    @Override
    public void onEnable() {
        if(mc.thePlayer == null)
            return;

        oldX = mc.thePlayer.posX;
        oldY = mc.thePlayer.posY;
        oldZ = mc.thePlayer.posZ;

        fakePlayer = new EntityOtherPlayerMP(mc.theWorld, mc.thePlayer.getGameProfile());
        fakePlayer.clonePlayer(mc.thePlayer, true);

        fakePlayer.rotationYawHead = mc.thePlayer.rotationYawHead;
        fakePlayer.copyLocationAndAnglesFrom(mc.thePlayer);

        mc.theWorld.addEntityToWorld(-1000, fakePlayer);

        if(noClipValue.get())
            mc.thePlayer.noClip = true;
    }

    @Override
    public void onDisable() {
        if(mc.thePlayer == null || fakePlayer == null)
            return;

        mc.thePlayer.setPositionAndRotation(oldX, oldY, oldZ, mc.thePlayer.rotationYaw, mc.thePlayer.rotationPitch);
        mc.theWorld.removeEntityFromWorld(fakePlayer.getEntityId());
        fakePlayer = null;
        mc.thePlayer.motionX = 0;
        mc.thePlayer.motionY = 0;
        mc.thePlayer.motionZ = 0;
    }

    @EventTarget
    public void onUpdate(UpdateEvent event) {
        if(noClipValue.get())
            mc.thePlayer.noClip = true;
        mc.thePlayer.fallDistance = 0;

        if(flyValue.get()) {
            final float value = speedValue.get();

            mc.thePlayer.motionY = 0;
            mc.thePlayer.motionX = 0;
            mc.thePlayer.motionZ = 0;
            if(mc.gameSettings.keyBindJump.isKeyDown())
                mc.thePlayer.motionY += value;
            if(mc.gameSettings.keyBindSneak.isKeyDown())
                mc.thePlayer.motionY -= value;
            MovementUtils.strafe(value);
        }
    }

    @EventTarget
    public void onPacket(PacketEvent event) {
        final Packet<?> packet = event.getPacket();

        if(packet instanceof C03PacketPlayer || packet instanceof C0BPacketEntityAction)
            event.cancelEvent();
    }
}

File: shared/main/java/net/ccbluex/liquidbounce/features/module/modules/render/FreeCam.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.render

import net.ccbluex.liquidbounce.event.EventTarget
import net.ccbluex.liquidbounce.event.PacketEvent
import net.ccbluex.liquidbounce.event.UpdateEvent
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.utils.FakePlayer
import net.ccbluex.liquidbounce.utils.extensions.strafe
import net.ccbluex.liquidbounce.utils.extensions.zeroXYZ
import net.ccbluex.liquidbounce.value.BoolValue
import net.ccbluex.liquidbounce.value.FloatValue

@ModuleInfo(name = "FreeCam", description = "Allows you to move out of your body.", category = ModuleCategory.RENDER)
class FreeCam : Module()
{
	private val speedValue = FloatValue("Speed", 0.8f, 0.1f, 2f)
	private val flyValue = BoolValue("Fly", true)
	private val noClipValue = BoolValue("NoClip", true)

	private var fakePlayer: FakePlayer? = null

	private var oldX = 0.0
	private var oldY = 0.0
	private var oldZ = 0.0
	private var oldYaw = 0.0f
	private var oldPitch = 0.0f
	private var oldGround = false

	override fun onEnable()
	{
		val thePlayer = mc.thePlayer ?: return
		val theWorld = mc.theWorld ?: return

		oldX = thePlayer.posX
		oldY = thePlayer.posY
		oldZ = thePlayer.posZ
		oldYaw = thePlayer.rotationYaw
		oldPitch = thePlayer.rotationPitch
		oldGround = thePlayer.onGround

		fakePlayer = FakePlayer(theWorld, thePlayer, -13370)

		if (noClipValue.get()) thePlayer.noClip = true
	}

	override fun onDisable()
	{
		val thePlayer = mc.thePlayer ?: return

		fakePlayer?.destroy()

		thePlayer.setPositionAndRotation(oldX, oldY, oldZ, oldYaw, oldPitch)

		thePlayer.zeroXYZ()
	}

	@EventTarget
	fun onUpdate(@Suppress("UNUSED_PARAMETER") event: UpdateEvent)
	{
		val thePlayer = mc.thePlayer ?: return

		if (noClipValue.get()) thePlayer.noClip = true

		thePlayer.fallDistance = 0.0f

		if (flyValue.get())
		{
			val value = speedValue.get()

			thePlayer.zeroXYZ()

			val gameSettings = mc.gameSettings
			if (gameSettings.keyBindJump.isKeyDown) thePlayer.motionY += value
			if (gameSettings.keyBindSneak.isKeyDown) thePlayer.motionY -= value

			thePlayer.strafe(value)
		}
	}

	@EventTarget
	fun onPacket(event: PacketEvent)
	{
		val packet = event.packet

		val provider = classProvider

		if (provider.isCPacketPlayer(packet)) // To bypass FreeCam checks, we need to keep sending normal packets.
		{
			val movePacket = packet.asCPacketPlayer()

			if (movePacket.moving)
			{
				movePacket.x = oldX
				movePacket.y = oldY
				movePacket.z = oldZ
			}

			movePacket.onGround = oldGround

			if (movePacket.rotating)
			{
				movePacket.yaw = oldYaw
				movePacket.pitch = oldPitch
			}
		}

		if (provider.isCPacketEntityAction(packet)) event.cancelEvent()
	}

	override val tag: String
		get() = "$oldX, $oldY, $oldZ, $oldGround, $oldYaw, $oldPitch"
}
